pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -254,31 +263,81 @@ var planMutators = []mutator{
 func (p *testPlanner) Plan() *TestPlan {
 	setup, testUpgrades := p.setupTest()
 
-	planUpgrade := func(upgrade *upgradePlan, virtualClusterSetup, scheduleHooks bool) {
-		for _, s := range p.services() {
-			s.startUpgrade(upgrade.to)
+	upgradeStepsForService := func(
+		service *ServiceContext, from, to *clusterupgrade.Version, virtualClusterSetup, scheduleHooks bool,
+	) []testStep {
+		if p.deploymentMode == SeparateProcessDeployment {
+			// Separate-process deployments are different because the two
+			// services involved (system and tenant) restart at different
+			// times (since they run separate binaries): the system (storage
+			// cluster) needs to upgrade to completion first, and only then
+			// the tenant is allowed to start upgrading.
+			//
+			// To allow for internal and external hooks to recognize this
+			// stage, we set update the service's stage accordingly below.
```

Comment:
nit: I understand what's going on now, but it took me a couple rereads to get why we set the tenant to `UpgradingSystem` and system to `UpgradingTenant`. Maybe a mention that this stage means we're waiting for the system/tenant to finish upgrading might've nudged me in the right direction faster.

Reply:
I updated the comments slightly here, hopefully it made it a little clearer.